datagen/src/main/scala/com/nvidia/rapids/tests/scaletest/TableGenerator.scala

```diff
@@ -0,0 +1,270 @@
+    // TODO: maybe we can move this to an external text file so we can extend it easily in the future
+    val candidates = Seq(
+      "int",
+      "long",
+      "Decimal(7, 2)",
+      "Decimal(19, 4)",
+      "string",
+      "map<string, struct<lat: Decimal(7,4), lon: Decimal(7, 4)>>",
+      "timestamp",
+      "date")
+    candidates(random.nextInt(candidates.length))
+  }
+
+  private def randomNumericColumnType(): String = {
+    val candidates = Seq(
+      "int",
+      "long",
+      "Decimal(7, 2)",
+      "Decimal(19, 4)"
+    )
+    candidates(random.nextInt(candidates.length))
+  }
+
+  /**
+  a_facts: Each scale factor corresponds to 10,000 rows
+    - primary_a the primary key from key group 1
+    - key group 4
+    - a_data_low_unique_1 data column that is a long with a unique count of 5
+    - a_data_low_unique_len_1 data column that is a string with variable length ranging from 1 to 5 inclusive (even distribution of lengths)
+    - complexity data columns
+   */
+  private def genAFacts(): DataFrame = {
+    val schema = "primary_a long," +
+      "a_key4_1 int," +
+      "a_data_low_unique_1 long," +
+      "a_data_low_unique_len_1 string," +
+      (1 to complexity).map(i => s"a_data_$i ${randomColumnType()}").mkString(",")
+    val aFacts = dbgen.addTable("a_facts", schema, aNumRows)
+    aFacts("primary_a").setSeedRange(1, aNumRows).setSeedMapping(DistinctDistribution())
+    aFacts("a_key4_1").setSeedRange(1,5)
+    aFacts("a_data_low_unique_1").setSeedRange(1, 5)
+    aFacts("a_data_low_unique_len_1").setLength(1, 5)
+    // ride-along columns will all use default data.
+    aFacts.toDF(spark)
+  }
+
+  /**
+  b_data: Each scale factor corresponds to 1,000,000 rows. (The row group size should be configured to be 512 MiB when writing parquet or equivalent when writing ORC)
+    - b_foreign_a Should overlap with a_facts.primary_a about 99% of the time
+    - key group 3 with 3 columns (The unique count should be about 99% of the total number of rows).
+    - 10 data columns of various types
+    - 10 decimal(10, 2) columns with a relatively small value range
+   */
+  private def genBData(): DataFrame = {
+    val schema = "b_foreign_a long," +
+      "b_key3_1 string," +
+      "b_key3_2 date," +
+      "b_key3_3 long," +
+      (1 to 10).map(i => s"b_data_$i ${randomColumnType()}").mkString(", ") + ", " +
+      (1 to 10).map(i => s"b_data_small_value_range_$i decimal(10, 2)").mkString(", ")
+    val overlapRatio = 0.99
+    val bData = dbgen.addTable("b_data", schema, bNumRows)
+    // This column "Should overlap with a_facts.primary_a about 99% of the time"
+    val overlapSeeds = Math.round(aNumRows * overlapRatio)
+    val startSeed = aNumRows - overlapSeeds + 1
+    val endSeed = startSeed + aNumRows
+    bData("b_foreign_a").setSeedRange(startSeed, endSeed)
+    bData.configureKeyGroup(
+      (1 to 3).map(i => s"b_key3_$i"),
+      CorrelatedKeyGroup(1, 1, Math.round(bNumRows * 0.99)),
```

Comment:
nit: can we have the id of the CorrelatedKeyGroup match the key id, so this would be 3?